pkg/destroy/alibabacloud/alibabacloud.go

```diff
@@ -0,0 +1,1380 @@
+	execute func() error
+}, errCh chan error, wg *sync.WaitGroup) error {
+	defer wg.Done()
+
+	err := wait.PollImmediateInfinite(
+		time.Second*10,
+		func() (bool, error) {
+			ferr := f.execute()
+			if ferr != nil {
+				o.Logger.Debugf("%s: %v", f.name, ferr)
+				return false, nil
+			}
+			return true, nil
+		},
+	)
+
+	if err != nil {
+		errCh <- err
+	}
+	return nil
+}
+
+func (o *ClusterUninstaller) findResources() (err error) {
+	tagResources, err := o.findResourcesByTag()
+	if err != nil {
+		return errors.Wrap(err, "failed to find resource by tag")
+	}
+	if len(tagResources) > 0 {
+		var deletedResources []ResourceArn
+		for _, resource := range tagResources {
+			arn, err := convertResourceArn(resource.ResourceARN)
+			if err != nil {
+				continue
+			}
+			deletedResources = append(deletedResources, arn)
+		}
+
+		o.tidyResourceByArn(deletedResources)
+	}
+	return nil
+}
+
+func (o *ClusterUninstaller) tidyResourceByArn(resourceArns []ResourceArn) {
+	for _, resourceArn := range resourceArns {
+		switch resourceArn.Service {
+		case "ecs":
+			switch resourceArn.ResourceType {
+			case "instance":
+				o.TagResources.ecsInstances = append(o.TagResources.ecsInstances, resourceArn)
+			case "securitygroup":
+				o.TagResources.securityGroups = append(o.TagResources.securityGroups, resourceArn)
+			default:
+				o.TagResources.others = append(o.TagResources.others, resourceArn)
+			}
+		case "vpc":
+			switch resourceArn.ResourceType {
+			case "vpc":
+				o.TagResources.vpcs = append(o.TagResources.vpcs, resourceArn)
+			case "vswitch":
+				o.TagResources.vSwitchs = append(o.TagResources.vSwitchs, resourceArn)
+			case "eip":
+				o.TagResources.eips = append(o.TagResources.eips, resourceArn)
+			case "natgateway":
+				o.TagResources.natgateways = append(o.TagResources.natgateways, resourceArn)
+			default:
+				o.TagResources.others = append(o.TagResources.others, resourceArn)
+			}
+		case "slb":
+			switch resourceArn.ResourceType {
+			case "instance":
+				o.TagResources.slbs = append(o.TagResources.slbs, resourceArn)
+			default:
+				o.TagResources.others = append(o.TagResources.others, resourceArn)
+			}
+		default:
+			o.TagResources.others = append(o.TagResources.others, resourceArn)
+		}
+	}
+}
+
+func (o *ClusterUninstaller) waitComplete() (err error) {
+	var isSuccessful bool
+	err = wait.Poll(
+		2*time.Second,
+		30*time.Second,
+		func() (bool, error) {
+			tagResources, err := o.findResourcesByTag()
+			if err != nil {
+				return false, err
+			}
+			if len(tagResources) == 0 {
+				isSuccessful = true
+				return true, nil
+			}
+			return false, nil
+		},
+	)
+
+	if !isSuccessful {
+		notDeletedResources := []string{}
+		tagResources, err := o.findResourcesByTag()
+		if err != nil {
+			return err
+		}
+		for _, arn := range tagResources {
+			notDeletedResources = append(notDeletedResources, arn.ResourceARN)
+		}
+		return errors.New(fmt.Sprintf("There are undeleted cloud resources %q", notDeletedResources))
```

Comment:
As staebler pointed out in the previous PR, the destroy code should never stop running if it knows that there are outstanding resources to be deleted. The destroyer should loop, attempting to destroy any resources as long as they exist.

Reply:
Yes, it is inappropriate to raise an error here.
But I think `waitComplete` here is redundant. The delete function of each cloud resource already contains the logic to wait for the deletion to be completed, so I should remove `waitComplete`.